presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,161 @@
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+
+@ThreadSafe
+public class AggregatedMemoryContext
+{
+    public static final long GUARANTEED_MEMORY = new DataSize(1, MEGABYTE).toBytes();
+
+    @Nullable
+    protected final AggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    private final MemoryReservationHandler reservationHandler;
+    @GuardedBy("this")
+    protected long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public AggregatedMemoryContext()
+    {
+        this(null, null);
+    }
+
+    public AggregatedMemoryContext(MemoryReservationHandler reservationHandler)
+    {
+        this(null, reservationHandler);
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler)
+    {
+        this.parentMemoryContext = parentMemoryContext;
+        this.reservationHandler = reservationHandler;
+    }
+
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this, null);
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new LocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    synchronized ListenableFuture<?> updateBytes(long bytes)
+    {
+        if (bytes == 0) {
+            return NOT_BLOCKED;
+        }
+
+        ListenableFuture<?> future = NOT_BLOCKED;
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
+            // update the parent before updating usedBytes as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+            future = parentMemoryContext.updateBytes(bytes);
+            usedBytes += bytes;
+            return future;
+        }
+
+        // reservationHandler is called only by the root AggregatedMemoryContext
+        if (reservationHandler != null) {
+            future = reservationHandler.reserveMemory(bytes);
+        }
+
+        usedBytes += bytes;
+
+        // Never block queries using a trivial amount of memory
+        if (usedBytes < GUARANTEED_MEMORY) {
+            future = NOT_BLOCKED;
+        }
+
+        return future;
+    }
+
+    synchronized boolean tryReserveMemory(long bytes)
+    {
+        if (bytes == 0) {
+            return true;
+        }
+
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
+            if (parentMemoryContext.tryReserveMemory(bytes)) {
+                usedBytes += bytes;
+                return true;
+            }
+            return false;
+        }
+        // reservationHandler is called only by the root AggregatedMemoryContext
+        if (reservationHandler != null) {
+            if (reservationHandler.tryReserveMemory(bytes)) {
+                usedBytes += bytes;
+                return true;
+            }
+            return false;
+        }
+        return true;
```

Comment:
This seems wrong. If neither parentMemoryContext nor reservationHandler exists, the function returns true and does not update `usedBytes`.